File: SIGame/Components.MSScriptControl/Bomb.js
var _x = 0;
var _y = 0;
var _bombSpeed = 0;
var _dispatcher = null;
var _alive = false;

function asign(dispatcher, x, y, bombSpeed) {
    _x = x;
    _y = y;
    _bombSpeed = bombSpeed;
    _dispatcher = dispatcher;
}

function drop() {
    _alive = true;
    var worker = new Worker(getScriptPath(function () {
        self.addEventListener('message', function (e) {
            var value = 0;
            while (_y <= _dispatcher.height && _alive) {
                _y++;
                await sleep(_bombSpeed);
            }
            _dispatcher.destroyBomb(this);
        }, false);
    }));
    worker.postMessage(0);
}

function destroy() {
    _alive = false;
}

function draw(screen) {
    screen.draw(_x, _y, "+");
}

File: Bomb/Bomb.cs
using SIGameLibrary;
using System;
using System.Drawing;
using System.Runtime.InteropServices;
using System.Threading;

namespace Bomb
{
    [ProgId("SIGame.Component.Bomb"), Guid("645A0CB1-1E57-4E65-A3D9-99F57BC30F4F"),
       ClassInterface(ClassInterfaceType.None), ComSourceInterfaces(typeof(IBomb))]
    [ComVisible(true)]
    public class Bomb : IBomb  // Бомба
    {
        private int _x;  // X координата
        private int _y;  // Y координата
        private bool _alive;  // Жив ли поток
        private int _delay;  // Задержка между действиями
        private Thread _thread;  // Поток
        private dynamic _dispatcher;  // Диспетчер

        public Point Position { get => new Point(_x, _y); }  // Позиция

        public dynamic Assign(params object[] parameters)  // Инициализация переменных
        {
            if (parameters.Length != 4)
                throw new ArgumentException("Ivalid argument number!");
            _dispatcher = parameters[0];
            _x = (int)parameters[1];
            _y = (int)parameters[2];
            _delay = (int)(150 / (float)parameters[3]);
            return this;
        }

        private void ThreadLoop()  // Цикл потока
        {
            while (_alive)  
            {
                if (_y <= _dispatcher.Height - 4)  
                    _y++;  
                else
                    _dispatcher.DestroyBomb(this);
                Thread.Sleep(_delay);  
            }
        }

        public void Drop()  // Скинуть бомбу
        {
            _alive = true;  
            _thread = new Thread(ThreadLoop);  
            _thread.Start();  
        }

        public void Destroy()  // Уничтожить бомбу
        {
            _alive = false;  
        }

        public void Draw(DBCharScreen screen)  // Нарисовать бомбу
        {
            screen.Draw(_x, _y, "+", ConsoleColor.Green);  
        }
    }
}
